main.go

```diff
@@ -42,6 +52,77 @@ type Output struct {
 	Expiration      string
 }
 
+type Cache struct {
+	Dir string
+}
+
+func NewCache(dir string) (*Cache, error) {
+	if _, err := os.Open(dir); err != nil {
+		return nil, err
+	}
+
+	fn := path.Join(dir, "validate")
+	if _, err := os.Create(fn); err != nil {
+		return nil, err
+	}
+	os.Remove(fn)
+
+	return &Cache{
+		Dir: dir,
+	}, nil
+}
+
+func (c *Cache) filenameByRole(role string) string {
+	h := sha256.New()
+	h.Write([]byte(role))
+	return fmt.Sprintf("%x", h.Sum(nil))
+}
+
+func (c *Cache) Get(role string) (*Output, bool) {
+	fp := path.Join(c.Dir, c.filenameByRole(role))
+	if f, err := os.Open(fp); err != nil {
+		return nil, false
+	} else {
+
+		if bytes, err := io.ReadAll(f); err != nil {
+			return nil, false
+		} else {
+			output := Output{}
+			if err := json.Unmarshal(bytes, &output); err != nil {
+				return nil, false
+			}
+
+			// check expiration
+			if tm, err := time.Parse(time.RFC3339, output.Expiration); err == nil && tm.After(time.Now().Add(REFRESH_HEAD_ROOM)) {
+				return &output, true
+			}
+
+			// don't use cache
+			return nil, false
+		}
+	}
+}
+
+func (c *Cache) Set(role string, output *Output) error {
+	fn := c.filenameByRole(role)
+	fp := path.Join(c.Dir, fn)
+	if bytes, err := json.Marshal(output); err != nil {
+		return err
+	} else {
+		if f, err := os.CreateTemp(c.Dir, fn+"-"); err != nil {
+			return err
+		} else {
+			f.Close()
+			if err := os.WriteFile(f.Name(), bytes, 0600); err != nil {
+				return err
+			}
+			os.Rename(f.Name(), fp)
+		}
+
+		return nil
+	}
```

Comment:
Suggestion to remove else.

```suggestion
	bytes, err := json.Marshal(output)
	if err != nil {
		return err
	}

	f, err := os.CreateTemp(c.Dir, fn+"-")
	if err != nil {
		return err
	}
	f.Close()
	
	if err := os.WriteFile(f.Name(), bytes, 0600); err != nil {
		return err
	}
	os.Rename(f.Name(), fp)
```